realm/realm-library/src/objectServer/java/io/realm/RealmApp.java

```diff
@@ -98,14 +85,95 @@ public RealmApp(String appId) {
+            networkTransport.addCustomRequestHeader(entry.getKey(), entry.getValue());
+        }
+        this.syncManager = new SyncManager(this);
+        this.nativePtr = init(config);
+    }
+
+    private long init(RealmAppConfiguration config) {
+        // Setup Realm part of User-Agent string
+        String userAgentBindingInfo = "Unknown"; // Fallback in case of anything going wrong
+        try {
+            StringBuilder sb = new StringBuilder();
+            sb.append("RealmJava/");
+            sb.append(BuildConfig.VERSION_NAME);
+            sb.append(" (");
+            sb.append(Util.isEmptyString(Build.DEVICE) ? "unknown-device" : Build.DEVICE);
+            sb.append(", ");
+            sb.append(Util.isEmptyString(Build.MODEL) ? "unknown-model" : Build.MODEL);
+            sb.append(", v");
+            sb.append(Build.VERSION.SDK_INT);
+            sb.append(")");
+            userAgentBindingInfo = sb.toString();
+        } catch (Exception e) {
+            // Failures to construct the user agent should never cause the system itself to crash.
+            RealmLog.warn("Constructing User-Agent description failed.", e);
+        }
+
+        // Create app UserAgent string
+        String appDefinedUserAgent = null;
+        String appName = config.getAppName();
+        String appVersion = config.getAppVersion();
+        if (!Util.isEmptyString(appName) || !Util.isEmptyString(appVersion)) {
+            StringBuilder sb = new StringBuilder();
+            sb.append(Util.isEmptyString(appName) ? "Undefined" : appName);
+            sb.append('/');
+            sb.append(Util.isEmptyString(appName) ? "Undefined" : appVersion);
+            appDefinedUserAgent = sb.toString();
+        }
+
+        if (BaseRealm.applicationContext == null) {
+            throw new IllegalStateException("Call Realm.init() first.");
+        }
+        Context context = BaseRealm.applicationContext;
+
+        String syncDir = null;
+        if (SyncManager.Debug.separatedDirForSyncManager) {
+            try {
+                // Files.createTempDirectory is not available on JDK 6.
+                File dir = File.createTempFile("remote_sync_", "_" + android.os.Process.myPid(), context.getFilesDir());
+                if (!dir.delete()) {
+                    throw new IllegalStateException(String.format(Locale.US,
+                            "Temp file '%s' cannot be deleted.", dir.getPath()));
+                }
+                if (!dir.mkdir()) {
+                    throw new IllegalStateException(String.format(Locale.US,
+                            "Directory '%s' for SyncManager cannot be created. ",
+                            dir.getPath()));
+                }
+                syncDir = dir.getPath();
+            } catch (IOException e) {
+                throw new IllegalStateException(e);
+            }
+        } else {
+            syncDir = context.getFilesDir().getPath();
+        }
+
```

Comment:
We could improve the readability in this method by hiding the computation of `syncDir`, `userAgentBindingInfo` and `appDefinedUserAgent` behind helper methods/classes. These could be tested in isolation making it less error-prone.

Reply:
Yes, I split the class into smaller methods